File: app/src/main/java/com/mytlogos/enterprise/ui/ImageViewerFragment.java
package com.mytlogos.enterprise.ui;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DecodeFormat;
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;
import com.bumptech.glide.request.target.Target;
import com.mytlogos.enterprise.R;
import com.mytlogos.enterprise.background.Repository;
import com.mytlogos.enterprise.background.RepositoryImpl;
import com.mytlogos.enterprise.model.ChapterPage;
import com.mytlogos.enterprise.model.SimpleEpisode;
import com.mytlogos.enterprise.tools.FileTools;
import com.mytlogos.enterprise.tools.ImageContentTool;

import java.io.File;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem;
import eu.davidea.flexibleadapter.items.IFlexible;
import eu.davidea.viewholders.FlexibleViewHolder;

public class ImageViewerFragment extends ViewerFragment<ImageViewerFragment.ReadableEpisode> {
    private FlexibleAdapter<IFlexible<ViewHolder>> adapter;


    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment ImageViewerFragment.
     */
    public static ImageViewerFragment newInstance(int startEpisode, String zipFile) {
        ImageViewerFragment fragment = new ImageViewerFragment();
        Bundle args = new Bundle();
        args.putInt(START_EPISODE, startEpisode);
        args.putString(MEDIUM, zipFile);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    int getLayoutRes() {
        return R.layout.image_reader_fragment;
    }

    @NonNull
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        RecyclerView recyclerView = view.findViewById(R.id.list);

        this.adapter = new FlexibleAdapter<>(null);
        LinearLayoutManager manager = new LinearLayoutManager(getContext());
        recyclerView.setLayoutManager(manager);

        DividerItemDecoration decoration = new DividerItemDecoration(requireContext(), manager.getOrientation());
        recyclerView.addItemDecoration(decoration);
        this.adapter.addListener((FlexibleAdapter.OnItemClickListener) (view1, position) -> {
            this.toggleReadingMode();
            return false;
        });

        recyclerView.setAdapter(this.adapter);
        recyclerView.setOnScrollChangeListener(
                (v, scrollX, scrollY, oldScrollX, oldScrollY) ->
                        this.onScroll(scrollX, scrollY, oldScrollX, oldScrollY)
        );
        this.loadEpisodes();
        return view;
    }

    @Override
    int getScrolledViewId() {
        return R.id.list;
    }

    @Override
    float getCurrentProgress() {
        return this.currentlyReading != null ? this.currentlyReading.getProgress() : 0;
    }

    @Override
    void saveProgress(float progress) {
        if (this.currentEpisode > 0) {
            RepositoryImpl.getInstance().updateProgress(this.currentEpisode, progress);
        }
    }

    @Override
    void onScroll(int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
        super.onScroll(scrollX, scrollY, oldScrollX, oldScrollY);
        if (this.currentlyReading == null) {
            return;
        }
        float progress = this.calculateProgressByScroll(scrollY, scrollX);
        this.updateProgress(progress);
    }

    @Override
    float calculateProgressByScroll(int scrollY, int scrollX) {
        int childCount = this.adapter.getRecyclerView().getChildCount();

        if (childCount == 0) {
            return 0;
        }
        List<View> visibleChildren = new ArrayList<>();

        for (int i = 0; i < childCount; i++) {
            View child = this.adapter.getRecyclerView().getChildAt(i);
            if (child == null) {
                continue;
            }
            int height = child.getHeight();

            if (height > 0) {
                visibleChildren.add(child);
            }
        }
        if (visibleChildren.isEmpty()) {
            return 0;
        }

        Rect lastVisibleRect = null;
        View lastVisibleView = null;

        for (int i = visibleChildren.size() - 1; i >= 0; i--) {
            Rect rect = new Rect();

            View view = visibleChildren.get(i);
            view.getLocalVisibleRect(rect);

            if (rect.bottom > 0) {
                lastVisibleRect = rect;
                lastVisibleView = view;
                break;
            }
        }

        if (lastVisibleRect == null || lastVisibleView.getHeight() <= 0) {
            return 0;
        }
        int lastVisiblePosition = -1;
        for (FlexibleViewHolder holder : this.adapter.getAllBoundViewHolders()) {
            if (holder.getFrontView() == lastVisibleView) {
                lastVisiblePosition = holder.getFlexibleAdapterPosition();
                break;
            }
        }
        int itemCount = this.adapter.getItemCount();
        int viewedItems = 0;
        if (lastVisiblePosition > 0) {
            viewedItems = lastVisiblePosition;
        }
        float viewedProgress = viewedItems / ((float) itemCount);
        // todo height may be negative (doc), but when?
        //  check how the case where it is negative should be treated
        float lastItemProgress = Math.abs(lastVisibleRect.height()) / ((float) lastVisibleView.getHeight() * itemCount);
        return viewedProgress + lastItemProgress;
    }

    @Override
    void updateContent() {
        if (this.currentlyReading != null) {
            this.currentEpisode = this.currentlyReading.getEpisodeId();

            List<IFlexible<ViewHolder>> flexibles = new ArrayList<>();

            for (int i = 1; i < this.currentlyReading.pageMap.size() + 1; i++) {
                ChapterPage page = this.currentlyReading.pageMap.get(i);

                ImageItem item;
                if (page == null) {
                    item = new ImageItem("", i);
                } else {
                    item = new ImageItem(page.getPath(), i);
                }
                flexibles.add(item);
            }
            this.adapter.updateDataSet(flexibles);

            if (this.currentlyReading.getPartialIndex() > 0) {
                setTitle(String.format("Episode %s.%s", this.currentlyReading.getTotalIndex(), this.currentlyReading.getPartialIndex()));
            } else {
                setTitle(String.format("Episode %s", this.currentlyReading.getTotalIndex()));
            }
        } else {
            // TODO: 06.08.2019 display an empty episode indicator?
            System.out.println("empty episode");
        }
        onLoadFinished();
    }


    private static class ImageItem extends AbstractFlexibleItem<ViewHolder> {
        private final String imagePath;
        private final int page;

        private ImageItem(String imagePath, int page) {
            this.imagePath = imagePath;
            this.page = page;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            ImageItem imageItem = (ImageItem) o;

            return Objects.equals(imagePath, imageItem.imagePath);
        }

        @Override
        public int hashCode() {
            return imagePath != null ? imagePath.hashCode() : 0;
        }

        @Override
        public int getLayoutRes() {
            return R.layout.image_item;
        }

        @Override
        public ViewHolder createViewHolder(View view, FlexibleAdapter<IFlexible> adapter) {
            return new ViewHolder(view, adapter);
        }

        @Override
        public void bindViewHolder(FlexibleAdapter<IFlexible> adapter, ViewHolder holder, int position, List<Object> payloads) {
            boolean empty = this.imagePath.isEmpty();

            holder.emptyText.setVisibility(empty ? View.VISIBLE : View.GONE);
            holder.imageView.setVisibility(empty ? View.GONE : View.VISIBLE);

            if (empty) {
                Glide.with(holder.itemView).clear(holder.imageView);
//                holder.imageView.setImageURI(null);
                holder.emptyText.setText(String.format("Page %s is missing", this.page));
            } else {
                holder.emptyText.setText(null);
                Glide
                        .with(holder.itemView)
                        .load(Uri.fromFile(new File(this.imagePath)))
                        .format(DecodeFormat.PREFER_ARGB_8888)
                        .override(Target.SIZE_ORIGINAL)
                        .into(holder.imageView);
//                holder.imageView.setImageURI(Uri.fromFile(new File(this.imagePath)));
            }
        }
    }

    private static class CropBitMap extends BitmapTransformation {
        @Override
        protected Bitmap transform(@NonNull BitmapPool pool, @NonNull Bitmap toTransform, int outWidth, int outHeight) {
            return null;
        }

        @Override
        public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {

        }
    }


    private static class ViewHolder extends FlexibleViewHolder {
        private final ImageView imageView;
        private final TextView emptyText;

        private ViewHolder(View view, FlexibleAdapter adapter) {
            super(view, adapter);
            this.emptyText = view.findViewById(R.id.empty_view);
            this.imageView = view.findViewById(R.id.image);
            this.emptyText.setOnClickListener(this);
            this.imageView.setOnClickListener(this);
        }
    }


    static class ReadableEpisode extends SimpleEpisode {
        @SuppressLint("UseSparseArrays")
        private final Map<Integer, ChapterPage> pageMap;

        ReadableEpisode(SimpleEpisode episode, Map<Integer, ChapterPage> pageMap) {
            super(episode.getEpisodeId(), episode.getTotalIndex(), episode.getPartialIndex(), episode.getProgress());
            this.pageMap = pageMap;
        }
    }

    private void loadEpisodes() {
        @SuppressLint("StaticFieldLeak")
        AsyncTask<Void, Void, Void> task = new AsyncTask<Void, Void, Void>() {

            @Override
            protected Void doInBackground(Void... voids) {
                ImageContentTool bookTool = FileTools.INSTANCE.getImageContentTool(getMainActivity().getApplication());
                Map<Integer, Set<ChapterPage>> chapterPages = bookTool.getEpisodePagePaths(currentBook);

                if (chapterPages == null || chapterPages.isEmpty()) {
                    return null;
                }
                Repository repository = RepositoryImpl.Companion.getInstance(getMainActivity().getApplication());

                List<SimpleEpisode> episodes = repository.getSimpleEpisodes(chapterPages.keySet());

                for (SimpleEpisode simpleEpisode : episodes) {
                    int episodeId = simpleEpisode.getEpisodeId();
                    Set<ChapterPage> pages = chapterPages.get(episodeId);

                    if (pages == null || pages.isEmpty()) {
                        System.err.println("Could not find file for episodeId: " + episodeId);
                        continue;
                    }

                    @SuppressLint("UseSparseArrays")
                    Map<Integer, ChapterPage> pageMap = new HashMap<>();
                    int max = 0;

                    for (ChapterPage page : pages) {
                        pageMap.put(page.getPage(), page);

                        if (max < page.getPage()) {
                            max = page.getPage();
                        }
                    }
                    for (int i = 1; i < max; i++) {
                        pageMap.putIfAbsent(i, new ChapterPage(episodeId, i, ""));
                    }
                    ReadableEpisode readableEpisode = new ReadableEpisode(simpleEpisode, pageMap);

                    if (episodeId == currentEpisode) {
                        currentlyReading = readableEpisode;
                    }

                    readableEpisodes.add(readableEpisode);
                }
                return null;
            }

            @SuppressLint("DefaultLocale")
            @Override
            protected void onPostExecute(Void data) {
                updateContent();
            }

            @Override
            protected void onCancelled() {
                showToastError("Could not load Book");
            }
        };
        task.execute();
    }

    private void showToastError(String s) {
        getMainActivity().runOnUiThread(() -> showToast(s));
    }
}

File: app/src/main/java/com/mytlogos/enterprise/ui/ImageViewerFragment.kt
package com.mytlogos.enterprise.ui

import android.annotation.SuppressLint
import android.graphics.Rect
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DecodeFormat
import com.bumptech.glide.request.target.Target
import com.mytlogos.enterprise.R
import com.mytlogos.enterprise.background.RepositoryImpl.Companion.instance
import com.mytlogos.enterprise.background.repository.EpisodeRepository
import com.mytlogos.enterprise.model.ChapterPage
import com.mytlogos.enterprise.model.SimpleEpisode
import com.mytlogos.enterprise.tools.getImageContentTool
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import java.io.File
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.abs

@ExperimentalCoroutinesApi
class ImageViewerFragment : ViewerFragment<ImageViewerFragment.ReadableEpisode?>() {
    private lateinit var adapter: ImageAdapter
    private lateinit var recyclerView: RecyclerView

    override val layoutRes: Int
        get() = R.layout.image_reader_fragment

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        val recyclerView: RecyclerView = view.findViewById(R.id.list) as RecyclerView

        val manager = LinearLayoutManager(context)
        recyclerView.layoutManager = manager
        val decoration = DividerItemDecoration(requireContext(), manager.orientation)
        recyclerView.addItemDecoration(decoration)

        adapter = ImageAdapter { toggleReadingMode() }
        recyclerView.adapter = adapter
        loadEpisodes()
        return view
    }

    override val scrolledViewId: Int
        get() = R.id.list

    override val currentProgress: Float
        get() = currentlyReading?.progress ?: 0.0f

    override fun saveProgress(progress: Float) {
        if (currentEpisode > 0) {
            instance.updateProgress(currentEpisode, progress)
        }
    }

    override fun onScroll(scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int) {
        super.onScroll(scrollX, scrollY, oldScrollX, oldScrollY)
        if (currentlyReading == null) {
            return
        }
        val progress = calculateProgressByScroll(scrollY, scrollX)
        updateProgress(progress)
    }

    override fun calculateProgressByScroll(scrollY: Int, scrollX: Int): Float {
        // may be called before ImageViewerFragment::onCreateView has finished, so ignore this
        if (!this::recyclerView.isInitialized) {
            return 0.0f
        }
        val childCount = this.recyclerView.childCount
        if (childCount == 0) {
            return 0.0f
        }
        this.recyclerView.childCount
        val visibleChildren: MutableList<View> = ArrayList()

        for (i in 0 until childCount) {
            val child = this.recyclerView.getChildAt(i) ?: continue
            val height = child.height

            if (height > 0) {
                visibleChildren.add(child)
            }
        }
        if (visibleChildren.isEmpty()) {
            return 0.0f
        }
        var lastVisibleRect: Rect? = null
        var lastVisibleView: View? = null

        for (i in visibleChildren.indices.reversed()) {
            val rect = Rect()
            val view = visibleChildren[i]
            view.getLocalVisibleRect(rect)

            if (rect.bottom > 0) {
                lastVisibleRect = rect
                lastVisibleView = view
                break
            }
        }
        if (lastVisibleRect == null || lastVisibleView!!.height <= 0) {
            return 0.0f
        }
        val viewHolder = recyclerView.findContainingViewHolder(lastVisibleView)
        val lastVisiblePosition = viewHolder?.bindingAdapterPosition ?: -1
        val itemCount = adapter.itemCount
        var viewedItems = 0

        if (lastVisiblePosition > 0) {
            viewedItems = lastVisiblePosition
        }
        val viewedProgress = viewedItems / itemCount.toFloat()
        // todo height may be negative (doc), but when?
        //  check how the case where it is negative should be treated
        val lastItemProgress =
            abs(lastVisibleRect.height()) / (lastVisibleView.height.toFloat() * itemCount)
        return viewedProgress + lastItemProgress
    }

    override fun updateContent() {
        val localCurrentlyReading = currentlyReading

        if (localCurrentlyReading != null) {
            currentEpisode = localCurrentlyReading.episodeId
            val flexibles: MutableList<String> = localCurrentlyReading
                .pageMap
                .map { (_, page) -> page.path }
                .toMutableList()

            adapter.setItems(flexibles)

            setTitle("Episode ${localCurrentlyReading.toIndexString()}")
        } else {
            // TODO: 06.08.2019 display an empty episode indicator?
            println("empty episode")
        }
        onLoadFinished()
    }

    private class ImageAdapter(val clickListener: () -> Unit) : RecyclerView.Adapter<ViewHolder>(), ItemPositionable<String> {
        private var items: List<String> = ArrayList()

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val root = LayoutInflater.from(parent.context).inflate(
                R.layout.image_item,
                parent,
                false
            )
            root.setOnClickListener { clickListener() }
            return ViewHolder(root)
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val imagePath = getItemAt(position)

            val empty = imagePath.isNullOrEmpty()
            holder.emptyText.visibility = if (empty) View.VISIBLE else View.GONE
            holder.imageView.visibility = if (empty) View.GONE else View.VISIBLE

            if (empty) {
                Glide.with(holder.itemView).clear(holder.imageView)
                holder.emptyText.text = "Page $position is missing"
            } else {
                holder.emptyText.text = null
                Glide
                    .with(holder.itemView)
                    .load(Uri.fromFile(File(imagePath!!)))
                    .format(DecodeFormat.PREFER_ARGB_8888)
                    .override(Target.SIZE_ORIGINAL)
                    .into(holder.imageView)
            }
        }

        @SuppressLint("NotifyDataSetChanged")
        fun setItems(items: List<String>) {
            this.items = items
            this.notifyDataSetChanged()
        }

        override fun getItemCount(): Int {
            return this.items.size
        }

        override fun getItemAt(position: Int): String? {
            if (position >= 0 && position < items.size) {
                return items[position]
            }
            return null
        }
    }

    private class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val imageView: ImageView = view.findViewById(R.id.image) as ImageView
        val emptyText: TextView = view.findViewById(R.id.empty_view) as TextView
    }

    class ReadableEpisode(
        episode: SimpleEpisode,
        val pageMap: Map<Int, ChapterPage>,
    ) : SimpleEpisode(episode.episodeId, episode.totalIndex, episode.partialIndex, episode.progress)

    private fun loadEpisodes() {
        lifecycleScope.launch {
            try {
                val bookTool = getImageContentTool(mainActivity.application)
                val chapterPages: Map<Int, Set<ChapterPage>> =
                    bookTool.getEpisodePagePaths(currentBook)

                if (chapterPages.isEmpty()) {
                    showToast("No Pages available")
                    return@launch
                }
                val repository = EpisodeRepository.getInstance(mainActivity.application)
                val episodes = repository.getSimpleEpisodes(chapterPages.keys)

                for (simpleEpisode in episodes) {
                    val episodeId = simpleEpisode.episodeId
                    val pages = chapterPages[episodeId]

                    if (pages == null || pages.isEmpty()) {
                        System.err.println("Could not find file for episodeId: $episodeId")
                        continue
                    }
                    val pageMap: MutableMap<Int, ChapterPage> = mutableMapOf()
                    var max = 0

                    for (page in pages) {
                        pageMap[page.page] = page
                        if (max < page.page) {
                            max = page.page
                        }
                    }
                    for (i in 1 until max) {
                        pageMap.putIfAbsent(i, ChapterPage(episodeId, i, ""))
                    }
                    val readableEpisode = ReadableEpisode(simpleEpisode, pageMap)

                    if (episodeId == currentEpisode) {
                        currentlyReading = readableEpisode
                    }
                    readableEpisodes.add(readableEpisode)
                }
            } catch (e: Throwable) {
                showToast("Could not load Book")
            }
            updateContent()
        }
    }

    companion object {
        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment ImageViewerFragment.
         */
        fun newInstance(startEpisode: Int, zipFile: String): ImageViewerFragment {
            val fragment = ImageViewerFragment()
            val args = Bundle()
            args.putInt(START_EPISODE, startEpisode)
            args.putString(MEDIUM, zipFile)
            fragment.arguments = args
            return fragment
        }
    }
}